pyro/distributions/polynomial.py

```diff
@@ -0,0 +1,139 @@
+
+    where :math:`x_n` is the :math:`n`th input, :math:`y_n` is the :math:`n`th output, and :math:`c_n\\in\\mathbb{R}`,
+    :math:`\\left{a^{(n)}_{r,k}\\in\\mathbb{R}\\right}` are learnable parameters that are the output of an
+    autoregressive NN inputting :math:`x_{\\prec n}={x_1,x_2,\\ldots,x_{n-1}}`.
+
+    Together with `TransformedDistribution` this provides a way to create richer variational approximations.
+
+    Example usage:
+
+    >>> from pyro.nn import AutoRegressiveNN
+    >>> input_dim = 10
+    >>> count_degree = 4
+    >>> count_sum = 3
+    >>> base_dist = dist.Normal(torch.zeros(input_dim), torch.ones(input_dim))
+    >>> arn = AutoRegressiveNN(input_dim, [input_dim*10], param_dims=[(count_degree + 1)*count_sum])
+    >>> flow = PolynomialFlow(arn, input_dim=input_dim, count_degree=count_degree, count_sum=count_sum)
+    >>> pyro.module("my_flow", flow)  # doctest: +SKIP
+    >>> flow_dist = dist.TransformedDistribution(base_dist, [flow])
+    >>> flow_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    The inverse of this transform does not possess an analytical solution and is left unimplemented. However,
+    the inverse is cached when the forward operation is called during sampling, and so samples drawn using
+    polynomial flow can be scored.
+
+    :param autoregressive_nn: an autoregressive neural network whose forward call returns a tensor of real-valued
+        numbers of size (batch_size, (count_degree+1)*count_sum, input_dim)
+    :type autoregressive_nn: nn.Module
+    :param count_degree: The degree of the polynomial to use for each element-wise transformation.
+    :type count_degree: int
+    :param count_sum: The number of polynomials to sum in each element-wise transformation.
+    :type count_sum: int
+
+    References:
+
+    Sum-of-squares polynomial flow. [arXiv:1905.02325]
+    Priyank Jaini, Kira A. Shelby, Yaoliang Yu
+    """
+
+    domain = constraints.real
+    codomain = constraints.real
+    bijective = True
+    event_dim = 1
+    autoregressive = True
+
+    def __init__(self, autoregressive_nn, input_dim, count_degree, count_sum):
+        super(PolynomialFlow, self).__init__(cache_size=1)
+
+        self.arn = autoregressive_nn
+        self.input_dim = input_dim
+        self.count_degree = count_degree
+        self.count_sum = count_sum
+        self._cached_logDetJ = None
+
+        self.c = nn.Parameter(torch.Tensor(input_dim))
+        self.reset_parameters()
+
+        # Vector of powers of input dimension
+        powers = torch.arange(1, count_degree + 2, dtype=torch.get_default_dtype())
+        self.register_buffer('powers', powers)
+
+        # Build mask of constants
+        mask = torch.zeros(count_degree + 1, count_degree + 1)
+        for jdx in range(count_degree + 1):
```

Comment:
can't this be replaced by something like
`mask = powers + torch.arange(count_degree + 1).unsqueeze(-1).type_as(powers)`

Reply:
Done!